proposals/nnnn-safe-shared-contiguous-storage.md

```diff
@@ -0,0 +1,1232 @@
+
+For example, for the hypothetical base64 decoding library mentioned above, a possible API could be:
+
+```swift
+extension HypotheticalBase64Decoder {
+  public func decode(bytes: some ContiguousStorage<UInt8>) -> [UInt8]
+}
+```
+
+Even better, an interface can be defined in terms of the concrete type `Span<UInt8>`:
+
+```swift
+extension Hypothetical Base64Decoder {
+  public func decode(bytes: Span<UInt8>) -> [UInt8]
+}
+```
+
+Advanced libraries might add use an inlinable generic-dispatch interface in addition to a concrete interface defined in terms of `Span`
+
+### `RawSpan`
+
+`RawSpan` allows sharing the contiguous internal representation for values which may be heterogenously-typed, such as in decoders. Since it is a fully concrete type, it can achieve better performance in debug builds of client code as well as a more straight-forwards understanding of performance for library code.
+
+`Span<some BitwiseCopyable>` can always be converted to `RawSpan`, using a conditionally-available property or a constructor.
+
+## Detailed design
+
+`Span<Element>` is a simple representation of a span of initialized memory.
+
+```swift
+public struct Span<Element: ~Copyable & ~Escapable>: Copyable, ~Escapable {
+  internal var _start: UnsafePointer<Element>
+  internal var _count: Int
+}
+```
+
+It provides a buffer-like interface to the elements stored in that span of memory:
+
+```swift
+extension Span {
+  public var count: Int { get }
+  public var isEmpty: Bool { get }
+  public var indices: Range<Int> { get }
+
+  subscript(_ position: Int) -> Element { get }
+}
+```
+
+Note that `Span` does _not_ conform to `Collection`. This is because `Collection`, as originally conceived and enshrined in existing source code, assumes pervasive copyability and escapability for itself as well as its elements. In particular a subsequence of a `Collection` is semantically a separate value from the instance it was derived from. In the case of `Span`, a sub-span representing a subrange of its elements _must_ have the same lifetime as the view from which it originates. Another proposal will consider collection-like protocols to accommodate different combinations of `~Copyable` and `~Escapable` for the collection and its elements.
+
+`Span`s representing subsets of consecutive elements can be extracted out of a larger `Span` with an API similar to the recently added `extracting()` functions of `UnsafeBufferPointer`:
+
+```swift
+extension Span {
+  public func extracting(_ bounds: Range<Int>) -> Self
+}
+```
+
+The first element of a given span is _always_ at position zero, and its last it as position `count-1`.
+
+As a side-effect of not conforming to `Collection` or `Sequence`, `Span` is not directly supported by `for` loops at this time. It is, however, easy to use in a `for` loop via indexing:
+
+```swift
+for i in mySpan.indices {
```

Comment:
I do not think we want to establish an expectation that noncopyable containers would come with a property called `indices`. Direct for-in loops over spans are expected to come very soon, in the proposal that introduces borrowing sequences. We do not need to bend over backwards to make this nice right now; I think we especially do not need to add any API to cater for it.

```suggestion
for i in 0 ..<  mySpan.count {
```

Reply:
Did this independently =)